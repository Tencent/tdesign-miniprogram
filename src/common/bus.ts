export default class Bus {
  listeners: Map<string, any>;

  constructor() {
    this.listeners = new Map();
  }

  on(evtName: string, listener) {
    const target = this.listeners.get(evtName) || [];

    target.push(listener);

    this.listeners.set(evtName, target);
  }

  emit(evtName: string) {
    const listeners = this.listeners.get(evtName);

    if (listeners) {
      listeners.forEach((func) => func());
    }
  }
}
