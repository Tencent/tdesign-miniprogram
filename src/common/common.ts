export type Classes = Array<string>;

export interface Styles {
  [css: string]: string | number;
}
/** 通用全局变量 */

export type OptionData = {
  label?: string;
  value?: string | number;
} & { [key: string]: any };

export type TreeOptionData = {
  children?: Array<TreeOptionData>;
} & OptionData;

export type SizeEnum = 'small' | 'medium' | 'large';

export type HorizontalAlignEnum = 'left' | 'center' | 'right';

export type VerticalAlignEnum = 'top' | 'middle' | 'bottom';

export type ClassName = { [className: string]: any } | ClassName[] | string;

export type CSSSelector = string;

export interface KeysType {
  value?: string;
  label?: string;
}
