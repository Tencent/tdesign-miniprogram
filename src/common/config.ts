export default {
  prefix: CONFIG_PREFIX,
};

export const prefix = CONFIG_PREFIX;
