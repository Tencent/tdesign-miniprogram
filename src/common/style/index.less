// base(含变量等)
@import './base.less';

// mixins
@import './mixins/_index.less';

// 工具类
@import './utilities/_index.less';
