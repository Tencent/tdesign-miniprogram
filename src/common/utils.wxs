/* utils */

/**
 * addUnit */
// 为 css 添加单位
function addUnit(value) {
  var REGEXP = getRegExp('^-?\d+(.\d+)?$');
  if (value == null) {
    return undefined;
  }
  return REGEXP.test('' + value) ? value + 'px' : value;
}

function isString(string) {
  return string && string.constructor === 'String';
}

function isArray(array) {
  return array && array.constructor === 'Array';
}

function isObject(obj) {
  return obj && obj.constructor === 'Object';
}

var isNoEmptyObj = function (obj) {
  return isObject(obj) && JSON.stringify(obj) !== '{}';
};

function includes(arr, value) {
  if (!arr || !isArray(arr)) return false;

  var i = 0;
  var len = arr.length;

  for (; i < len; i++) {
    if (arr[i] === value) return true;
  }
  return false;
}

function cls(base, arr) {
  var res = [base];
  var i = 0;
  for (var size = arr.length; i < size; i++) {
    var item = arr[i];

    if (item && item.constructor === 'Array') {
      var key = arr[i][0];
      var value = arr[i][1];

      if (value) {
        res.push(base + '--' + key);
      }
    } else if (typeof item === 'string' || typeof item === 'number') {
      if (item) {
        res.push(base + '--' + item);
      }
    }
  }
  return res.join(' ');
}

function getBadgeAriaLabel(options) {
  var maxCount = options.maxCount || 99;
  if (options.dot) {
    return '有新的消息';
  }
  if (options.count === '...') {
    return '有很多消息';
  }
  if (isNaN(options.count)) {
    return options.count;
  }
  var str1 = '有' + maxCount + '+条消息';
  var str2 = '有' + options.count + '条消息';
  return Number(options.count) > maxCount ? str1 : str2;
}

function endsWith(str, endStr) {
  return str.slice(-endStr.length) === endStr ? str : str + endStr;
}

function keys(obj) {
  return JSON.stringify(obj)
    .replace(getRegExp('{|}|"', 'g'), '')
    .split(',')
    .map(function (item) {
      return item.split(':')[0];
    });
}

function kebabCase(str) {
  return str
    .replace(getRegExp('[A-Z]', 'g'), function (ele) {
      return '-' + ele;
    })
    .toLowerCase();
}

function _style(styles) {
  if (isArray(styles)) {
    return styles
      .filter(function (item) {
        return item != null && item !== '';
      })
      .map(function (item) {
        return isArray(item) ? _style(item) : endsWith(item, ';');
      })
      .join(' ');
  }

  if (isObject(styles)) {
    return keys(styles)
      .filter(function (key) {
        return styles[key] != null && styles[key] !== '';
      })
      .map(function (key) {
        return [kebabCase(key), [styles[key]]].join(':');
      })
      .join(';');
  }

  return styles;
}

module.exports = {
  addUnit: addUnit,
  isString: isString,
  isArray: isArray,
  isObject: isObject,
  isNoEmptyObj: isNoEmptyObj,
  includes: includes,
  cls: cls,
  getBadgeAriaLabel: getBadgeAriaLabel,
  _style: _style,
};
