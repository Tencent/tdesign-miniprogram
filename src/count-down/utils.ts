export interface TimeData {
  DD: number;
  HH: number;
  mm: number;
  ss: number;
  SSS: number;
}
export const TimeDataUnit = {
  DD: '天',
  HH: '时',
  mm: '分',
  ss: '秒',
  SSS: '毫秒',
};

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

export const parseTimeData = function (time: number): TimeData {
  const days = Math.floor(time / DAY);
  const hours = Math.floor((time % DAY) / HOUR);
  const minutes = Math.floor((time % HOUR) / MINUTE);
  const seconds = Math.floor((time % MINUTE) / SECOND);
  const milliseconds = Math.floor(time % SECOND);
  return {
    DD: days,
    HH: hours,
    mm: minutes,
    ss: seconds,
    SSS: milliseconds,
  };
};

export const isSameSecond = function (time1: number, time2: number): boolean {
  return Math.floor(time1 / 1000) === Math.floor(time2 / 1000);
};

export type TTimeList = { digit: string; unit: string; match: string }[];

/**
 *
 * @param time 倒计时时间，毫秒单位
 * @param format 倒计时格式化字符串，例如：dd天hh小时mm分ss秒SSS毫秒，hh:mm:ss.SSS，hh:mm:ss
 */
export const parseFormat = function (time: number, format: string) {
  const obj = {
    'D+': Math.floor(time / 86400000), // 日
    'H+': Math.floor((time % 86400000) / 3600000), // 小时
    'm+': Math.floor((time % 3600000) / 60000), // 分
    's+': Math.floor((time % 60000) / 1000), // 秒
    'S+': Math.floor(time % 1000), // 毫秒
  };
  const timeList: TTimeList = [];
  let timeText = format;

  Object.keys(obj).forEach((prop) => {
    if (new RegExp(`(${prop})`).test(timeText)) {
      timeText = timeText.replace(RegExp.$1, (match, offset, source) => {
        const v = `${(obj as any)[prop]}`;
        let digit = v;
        if (match.length > 1) {
          digit = (match.replace(new RegExp(match[0], 'g'), '0') + v).substr(v.length);
        }
        const unit = source.substr(offset + match.length);
        const last = timeList[timeList.length - 1];
        if (last) {
          const index = last.unit.indexOf(match);
          if (index !== -1) {
            last.unit = last.unit.substr(0, index);
          }
        }
        timeList.push({ digit, unit, match });
        return digit;
      });
    }
  });
  return { timeText, timeList };
};
