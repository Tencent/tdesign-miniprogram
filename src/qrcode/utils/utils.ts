import type {
  ERROR_LEVEL_MAPPED_TYPE,
  ErrorCorrectionLevel,
  Excavation,
  ImageSettings,
  Modules,
} from './type';
import { Ecc } from './qrcodegen';

// =================== ERROR_LEVEL ==========================
export const ERROR_LEVEL_MAP: ERROR_LEVEL_MAPPED_TYPE = {
  L: Ecc.LOW,
  M: Ecc.MEDIUM,
  Q: Ecc.QUARTILE,
  H: Ecc.HIGH,
} as const;

// =================== DEFAULT_VALUE ==========================
export const DEFAULT_SIZE = 160;
export const DEFAULT_LEVEL: ErrorCorrectionLevel = 'M';
export const DEFAULT_BACKGROUND_COLOR = '#FFFFFF';
export const DEFAULT_FRONT_COLOR = '#000000';
export const DEFAULT_NEED_MARGIN = true;
export const DEFAULT_MINVERSION = 1;
export const SPEC_MARGIN_SIZE = 0;
export const DEFAULT_MARGIN_SIZE = 6; // 不同大小的二维码要修改默认margin
export const DEFAULT_IMG_SCALE = 0.1;

// =================== UTILS ==========================
/**
 * Generate a path string from modules
 * @param modules
 * @param margin
 * @returns
 */
export const generatePath = (modules: Modules, margin: number = 0) => {
  const ops: string[] = [];
  modules.forEach((row, y) => {
    let start: number | null = null;
    row.forEach((cell, x) => {
      if (!cell && start !== null) {
        ops.push(
          `M${start + margin} ${y + margin}h${x - start}v1H${start + margin}z`
        );
        start = null;
        return;
      }

      if (x === row.length - 1) {
        if (!cell) {
          return;
        }
        if (start === null) {
          ops.push(`M${x + margin},${y + margin} h1v1H${x + margin}z`);
        } else {
          ops.push(
            `M${start + margin},${y + margin} h${x + 1 - start}v1H${
              start + margin
            }z`
          );
        }
        return;
      }

      if (cell && start === null) {
        start = x;
      }
    });
  });
  return ops.join('');
};

/**
 * Excavate modules
 * @param modules
 * @param excavation
 * @returns
 */
export const excavateModules = (modules: Modules, excavation: Excavation) => modules.slice().map((row: any, y: any) => {
  if (y < excavation.y || y >= excavation.y + excavation.h) {
    return row;
  }
  return row.map((cell: any, x: any) => {
    if (x < excavation.x || x >= excavation.x + excavation.w) {
      return cell;
    }
    return false;
  });
});

/**
 * Get image settings
 * @param cells The modules of the QR code
 * @param size The size of the QR code
 * @param margin
 * @param imageSettings
 * @returns
 */
export const getImageSettings = (
  cells: Modules,
  size: number,
  margin: number,
  imageSettings?: ImageSettings
): null | {
  x: number;
  y: number;
  h: number;
  w: number;
  excavation: Excavation | null;
  opacity: number;
} => {
  if (imageSettings == null) {
    return null;
  }
  const numCells = cells.length + margin * 2;
  const defaultSize = Math.floor(size * DEFAULT_IMG_SCALE);
  const scale = numCells / size;
  const w = (imageSettings.width || defaultSize) * scale;
  const h = (imageSettings.height || defaultSize) * scale;
  const x = imageSettings.x == null
    ? cells.length / 2 - w / 2
    : imageSettings.x * scale;
  const y = imageSettings.y == null
    ? cells.length / 2 - h / 2
    : imageSettings.y * scale;
  const opacity = imageSettings.opacity == null ? 1 : imageSettings.opacity;

  let excavation = null;
  if (imageSettings.excavate) {
    const floorX = Math.floor(x);
    const floorY = Math.floor(y);
    const ceilW = Math.ceil(w + x - floorX);
    const ceilH = Math.ceil(h + y - floorY);
    excavation = { x: floorX, y: floorY, w: ceilW, h: ceilH };
  }

  return { x, y, h, w, excavation, opacity };
};

/**
 * Get margin size
 * @param needMargin Whether need margin
 * @param marginSize Custom margin size
 * @returns
 */
export const getMarginSize = (needMargin: boolean, marginSize?: number) => {
  if (marginSize != null) {
    return Math.max(Math.floor(marginSize), 0);
  }
  return needMargin ? SPEC_MARGIN_SIZE : DEFAULT_MARGIN_SIZE;
};

/**
 * Check if Path2D is supported
 */
export const isSupportPath2d = (() => {
  return false;
})();
