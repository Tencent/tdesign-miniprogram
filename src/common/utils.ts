import { prefix } from './config';

const systemInfo = wx.getSystemInfoSync();

type Context = WechatMiniprogram.Page.TrivialInstance | WechatMiniprogram.Component.TrivialInstance;

export const debounce = function (func, wait = 500) {
  let timerId;
  return function (...rest) {
    if (timerId) {
      clearTimeout(timerId);
    }
    timerId = setTimeout(() => {
      func.apply(this, rest);
    }, wait);
  };
};

export const throttle = (func, wait = 100, options = null) => {
  let previous = 0;
  let timerid = null;

  if (!options) {
    options = {
      leading: true,
    };
  }

  return function (...args) {
    const now = Date.now();

    if (!previous && !options.leading) previous = now;

    const remaining = wait - (now - previous);
    const context = this;

    if (remaining <= 0) {
      if (timerid) {
        clearTimeout(timerid);
        timerid = null;
      }
      previous = now;
      func.apply(context, args);
    }
  };
};

export const classNames = function (...args) {
  const hasOwn = {}.hasOwnProperty;
  const classes = [];

  args.forEach((arg) => {
    // for (let i = 0; i < args.length; i++) {
    // eslint-disable-next-line
    // const arg = args[i]
    if (!arg) return;

    const argType = typeof arg;

    if (argType === 'string' || argType === 'number') {
      classes.push(arg);
    } else if (Array.isArray(arg) && arg.length) {
      const inner = classNames(...arg);
      if (inner) {
        classes.push(inner);
      }
    } else if (argType === 'object') {
      // eslint-disable-next-line
      for (const key in arg) {
        if (hasOwn.call(arg, key) && arg[key]) {
          classes.push(key);
        }
      }
    }
  });

  return classes.join(' ');
};

export const styles = function (styleObj) {
  return Object.keys(styleObj)
    .map((styleKey) => `${styleKey}: ${styleObj[styleKey]}`)
    .join('; ');
};

export const getAnimationFrame = function (context: any, cb: Function) {
  return wx
    .createSelectorQuery()
    .in(context)
    .selectViewport()
    .boundingClientRect()
    .exec(() => {
      cb();
    });
};

export const getRect = function (context: any, selector: string, needAll: boolean = false) {
  return new Promise<any>((resolve, reject) => {
    wx.createSelectorQuery()
      .in(context)
      [needAll ? 'selectAll' : 'select'](selector)
      .boundingClientRect((rect) => {
        if (rect) {
          resolve(rect);
        } else {
          reject(rect);
        }
      })
      .exec();
  });
};

const isDef = function (value: any): boolean {
  return value !== undefined && value !== null;
};

export const isNumber = function (value) {
  return /^\d+(\.\d+)?$/.test(value);
};

export const addUnit = function (value?: string | number): string | undefined {
  if (!isDef(value)) {
    return undefined;
  }
  value = String(value);
  return isNumber(value) ? `${value}px` : value;
};

/**
 * 计算字符串字符的长度并可以截取字符串。
 * @param str 传入字符串（maxcharacter条件下，一个汉字表示两个字符）
 * @param max 规定最大字符串长度
 * @returns 当没有传入maxCharacter/maxLength 时返回字符串字符长度，当传入maxCharacter/maxLength时返回截取之后的字符串和长度。
 */
export const getCharacterLength = (type: string, str: string, max?: number) => {
  if (!str || str.length === 0) {
    return {
      length: 0,
      characters: '',
    };
  }

  if (type === 'maxcharacter') {
    let len = 0;
    for (let i = 0; i < str.length; i += 1) {
      let currentStringLength = 0;
      if (str.charCodeAt(i) > 127 || str.charCodeAt(i) === 94) {
        currentStringLength = 2;
      } else {
        currentStringLength = 1;
      }
      if (len + currentStringLength > max) {
        return {
          length: len,
          characters: str.slice(0, i),
        };
      }
      len += currentStringLength;
    }
    return {
      length: len,
      characters: str,
    };
  } else if (type === 'maxlength') {
    const length = str.length > max ? max : str.length;
    return {
      length,
      characters: str.slice(0, length),
    };
  }

  return {
    length: str.length,
    characters: str,
  };
};

export const chunk = (arr: any[], size: number) =>
  Array.from({ length: Math.ceil(arr.length / size) }, (v, i) => arr.slice(i * size, i * size + size));

export const getInstance = function (context?: Context, selector?: string) {
  if (!context) {
    const pages = getCurrentPages();
    const page = pages[pages.length - 1];
    context = page.$$basePage || page;
  }
  const instance = context ? context.selectComponent(selector) : null;
  if (!instance) {
    console.warn('未找到组件,请检查selector是否正确');
    return null;
  }
  return instance;
};

export const unitConvert = (value: number | string): number => {
  if (typeof value === 'string') {
    if (value.includes('rpx')) {
      return (parseInt(value, 10) * (systemInfo?.screenWidth ?? 750)) / 750;
    }
    return parseInt(value, 10);
  }
  return value;
};

export const setIcon = (iconName, icon, defaultIcon) => {
  if (icon) {
    if (typeof icon === 'string') {
      return {
        [`${iconName}Name`]: icon,
        [`${iconName}Data`]: {},
      };
    } else if (typeof icon === 'object') {
      return {
        [`${iconName}Name`]: '',
        [`${iconName}Data`]: icon,
      };
    } else {
      return {
        [`${iconName}Name`]: defaultIcon,
        [`${iconName}Data`]: {},
      };
    }
  }
  return {
    [`${iconName}Name`]: '',
    [`${iconName}Data`]: {},
  };
};

export const isBool = (val) => typeof val === 'boolean';

export const isObject = (val) => typeof val === 'object' && val != null;

export const isString = (val) => typeof val === 'string';

export const toCamel = (str) => str.replace(/-(\w)/g, (match, m1) => m1.toUpperCase());

export const getCurrentPage = function <T>() {
  const pages = getCurrentPages();
  return pages[pages.length - 1] as T & WechatMiniprogram.Page.TrivialInstance;
};

export const uniqueFactory = (compName) => {
  let number = 0;
  return () => `${prefix}_${compName}_${number++}`;
};

export const calcIcon = (icon: string | Record<string, any>, defaultIcon?: string) => {
  if ((isBool(icon) && icon && defaultIcon) || isString(icon)) {
    return { name: isBool(icon) ? defaultIcon : icon };
  }
  if (isObject(icon)) {
    return icon;
  }
  return null;
};

export const isOverSize = (size, sizeLimit) => {
  if (!sizeLimit) return false;

  const base = 1000;
  const unitMap = {
    B: 1,
    KB: base,
    MB: base * base,
    GB: base * base * base,
  };
  const computedSize =
    typeof sizeLimit === 'number' ? sizeLimit * base : sizeLimit?.size * unitMap[sizeLimit?.unit ?? 'KB']; // 单位 KB

  return size > computedSize;
};
