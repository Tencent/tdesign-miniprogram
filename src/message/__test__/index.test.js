import path from 'path';
import simulate from 'miniprogram-simulate';
import Message from '../index';
import * as Util from '../../common/utils';

const mockGetRect = jest.spyOn(Util, 'getRect');
mockGetRect.mockImplementation(() => {
  return new Promise((resolve) => resolve({ height: 46, width: 156 }));
});

const mockInstance = jest.spyOn(Util, 'getInstance');

describe('message', () => {
  const message = load(path.resolve(__dirname, `../message`), 't-message');

  jest.resetModules();
  const icon = load(path.resolve(__dirname, `../../icon/icon`), 't-icon');

  describe('props', () => {
    it(`: style && customStyle`, async () => {
      const id = simulate.load({
        template: `<t-message id="t-message" ></t-message>`,
        usingComponents: {
          't-message': message,
        },
      });
      const comp = simulate.render(id);
      comp.attach(document.createElement('parent-wrapper'));

      const $message = comp.querySelector('#t-message');

      mockInstance.mockImplementation(() => $message.instance);

      const style = 'color: red';
      const customStyle = 'font-size: 9px';

      Message.warning({
        context: $message.instance,
        offset: [20, 32],
        content: '这是一条带关闭的消息通知',
        duration: 0,
        style: style,
        customStyle: customStyle,
        closeBtn: true,
      });
      await simulate.sleep(540);

      expect(comp.toJSON()).toMatchSnapshot();

      const $style = comp.querySelector('#t-message >>> .t-message');
      if (VIRTUAL_HOST) {
        expect($style.dom.getAttribute('style').includes(`${style}; ${customStyle}`)).toBeTruthy();
      } else {
        expect($style.dom.getAttribute('style').includes(`${customStyle}`)).toBeTruthy();
      }
    });

    it(': icon', async () => {
      const id = simulate.load({
        template: `<t-message id="t-message" />`,
        usingComponents: {
          't-message': message,
        },
      });
      const comp = simulate.render(id);
      comp.attach(document.createElement('parent-wrapper'));
      const $message = comp.querySelector('#t-message');

      mockInstance
        .mockImplementation(() => $message.instance)
        .mockImplementationOnce(() => null)
        .mockImplementationOnce(() => null);

      // null: 控制台输出 console.error('未找到组件,请确认 selector && context 是否正确');
      Message.info({
        context: $message.instance,
      });
      await simulate.sleep();
      expect(comp.querySelector('#t-message >>> .t-message')).toBeUndefined();

      // null hide()方法未被调用
      Message.hide();

      Message.info({
        context: $message.instance,
        offset: [20, 32],
        icon: 'notification',
        marquee: {},
        content: 'iconName 为 notification',
        duration: -1,
      });
      await simulate.sleep(540);
      expect(comp.querySelector('#t-message >>> .t-message')).not.toBeUndefined();

      // theme
      const $theme = comp.querySelector('#t-message >>> .t-message');
      expect($theme.dom.getAttribute('class').includes('t-message--info')).toBe(true);

      // icon
      const $prefixIcon = comp.querySelector('#t-message >>> .t-message__icon--left');
      const iconId = simulate.load({
        template: `<t-icon id="t-icon" name="{{name}}"></t-icon>`,
        data: {
          name: 'notification',
        },
        usingComponents: {
          't-icon': icon,
        },
      });
      const iconComp = simulate.render(iconId).querySelector('#t-icon');
      expect($prefixIcon.dom.innerHTML).toContain(iconComp.dom.innerHTML);

      // content
      const $content = comp.querySelector('#t-message >>> .t-message__text');
      expect($content.dom.textContent).toContain('iconName 为 notification');
      comp.detach();
    });

    it(': marquee', async () => {
      const id = simulate.load({
        template: `<t-message id="t-message" />`,
        usingComponents: {
          't-message': message,
        },
      });
      const comp = simulate.render(id);
      comp.attach(document.createElement('parent-wrapper'));
      const $message = comp.querySelector('#t-message');

      mockInstance.mockImplementation(() => $message.instance);
      Message.info({
        context: $message.instance,
        offset: [20, 32],
        marquee: { speed: 100, loop: 3, delay: 0 },
        icon: false,
        content: 'icon 使用空值',
        duration: 0,
      });
      await simulate.sleep(540);
      // icon 为空
      const $prefixIcon = comp.querySelector('#t-message >>> .t-message__icon--left');
      expect($prefixIcon.dom.innerHTML).toBe('');
      Message.hide();
      await simulate.sleep(500);
      expect($message.instance.data.visible).toBe(false);
    });
  });

  describe('event', () => {
    it(': closeBtnClick', async () => {
      const actionBtnClick = jest.fn();
      const id = simulate.load({
        template: `<t-message id="t-message" bind:close-btn-click="actionBtnClick"/>`,
        usingComponents: {
          't-message': message,
        },
        methods: {
          actionBtnClick,
        },
      });
      const comp = simulate.render(id);
      comp.attach(document.createElement('parent-wrapper'));
      const $message = comp.querySelector('#t-message');

      mockInstance.mockImplementation(() => $message.instance);

      Message.warning({
        context: $message.instance,
        offset: [20, 32],
        content: '这是一条带关闭的消息通知',
        duration: 0,
        closeBtn: true,
      });
      await simulate.sleep(540);

      const $icon = comp.querySelector('#t-message >>> .t-message__icon--right');
      $icon.dispatchEvent('tap');
      await simulate.sleep(10);
      expect(actionBtnClick).toHaveBeenCalledTimes(1);
      Message.hide();
    });
  });
});
