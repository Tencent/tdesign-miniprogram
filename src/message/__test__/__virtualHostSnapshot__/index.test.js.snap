// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message props : style && customStyle 1`] = `
<main>
  <t-message
    id="t-message"
  >
    <wx-view
      animation="{{
        Object {
          "actions": Array [
            Object {
              "animates": Array [
                Object {
                  "args": Array [
                    66,
                  ],
                  "type": "translateY",
                },
              ],
              "option": Object {
                "transformOrigin": "50% 50% 0",
                "transition": Object {
                  "delay": 0,
                  "duration": 500,
                  "timingFunction": "ease",
                },
              },
            },
          ],
        }
      }}"
      ariaRole="alert"
      class="t-message class t-class t-message--warning"
      id="t-message"
      style="z-index:15000;top:-92rpx;right:32rpx;left:32rpx; color: red; font-size: 9px;"
    >
      <wx-view
        class="t-message__icon--left"
      >
        <t-icon
          class=""
          tClass="t-class-icon"
          bind:click=""
        >
          <wx-view
            ariaHidden="{{true}}"
            ariaLabel=""
            ariaRole=""
            class="t-icon class t-class"
            style=""
            bind:tap="onTap"
          >
            <wx-label
              class="t-icon-info-circle-filled t-icon-base"
            />
          </wx-view>
        </t-icon>
      </wx-view>
      <wx-view
        class="t-message__text-wrap "
        id="t-message__text-wrap"
        style="text-align: left"
      >
        <wx-view
          animation="{{Array []}}"
          class="t-message__text t-class-content"
          id="t-message__text"
        >
          这是一条带关闭的消息通知
        </wx-view>
      </wx-view>
      <wx-view
        class="t-message__icon--right"
        bind:tap="handleClose"
      >
        <t-icon
          class=""
          tClass="t-class-close-btn"
          bind:click=""
        >
          <wx-view
            ariaHidden="{{false}}"
            ariaLabel="关闭"
            ariaRole="button"
            class="t-icon class t-class"
            style=""
            bind:tap="onTap"
          >
            <wx-label
              class="t-icon-close t-icon-base"
            />
          </wx-view>
        </t-icon>
      </wx-view>
    </wx-view>
  </t-message>
</main>
`;
